src/main.rs

```diff
@@ -1,205 +1,337 @@
+        });
+        let message = self.message.format(now.into());
+        let checksum = self.crc.unwrap_or_else(|| checksum(message.as_bytes()));
+        format!("{message}*{checksum:X}\r\n").into_bytes()
+    }
 }
 
 fn checksum(buf: &[u8]) -> u8 {
-    let mut sum: u8 = 0;
+    let mut sum = 0;
     for c in &buf[1..] {
         sum ^= c;
     }
     sum
 }
 
-fn main() -> Result<()> {
-    let opt = Cli::parse();
+#[derive(Debug, Clone, Copy, serde::Deserialize)]
+#[serde(rename_all = "lowercase")]
+enum Message {
+    Gga {
+        lat: f64,
+        lon: f64,
+        height: f64,
+    },
+    Cra {
+        request_counter: Option<u8>,
+        area_id: Option<u32>,
+        corrections_mask: Option<u16>,
+        solution_id: Option<u8>,
+    },
+}
+
+impl Message {
+    fn format(&self, time: DateTime<Utc>) -> String {
+        match *self {
+            Message::Gga { lat, lon, height } => {
+                let time = time.format("%H%M%S.00");
+
+                let latn = ((lat * 1e8).round() / 1e8).abs();
+                let lonn = ((lon * 1e8).round() / 1e8).abs();
 
-    let latf: f64 = opt.lat.parse::<f64>()?;
-    let lonf: f64 = opt.lon.parse::<f64>()?;
-    let heightf: f64 = opt.height.parse::<f64>()?;
+                let lat_deg = latn as u16;
+                let lon_deg = lonn as u16;
+
+                let lat_min = (latn - (lat_deg as f64)) * 60.0;
+                let lon_min = (lonn - (lon_deg as f64)) * 60.0;
+
+                let lat_dir = if lat < 0.0 { 'S' } else { 'N' };
+                let lon_dir = if lon < 0.0 { 'W' } else { 'E' };
+
+                format!(
+                    "$GPGGA,{},{:02}{:010.7},{},{:03}{:010.7},{},4,12,1.3,{:.2},M,0.0,M,1.7,0078",
+                    time, lat_deg, lat_min, lat_dir, lon_deg, lon_min, lon_dir, height
+                )
+            }
+            Message::Cra {
+                request_counter,
+                area_id,
+                corrections_mask,
+                solution_id,
+            } => {
+                let mut s = String::from("$PSWTCRA,");
+                if let Some(request_counter) = request_counter {
+                    write!(&mut s, "{request_counter}").unwrap();
+                }
+                s.push(',');
+                if let Some(area_id) = area_id {
+                    write!(&mut s, "{area_id}").unwrap();
+                }
+                s.push(',');
+                if let Some(corrections_mask) = corrections_mask {
+                    write!(&mut s, "{corrections_mask}").unwrap();
+                }
+                s.push(',');
+                if let Some(solution_id) = solution_id {
+                    write!(&mut s, "{solution_id}").unwrap();
+                }
+                s
+            }
+        }
+    }
+}
+
+fn get_commands(opt: Cli) -> Result<Box<dyn Iterator<Item = Command>>> {
+    if opt.gga_period == 0 && opt.input.is_none() {
+        return Ok(Box::new(iter::empty()));
+    }
```

Comment:
I think you need to either update the doc comment on `gga_period` in the `CLI` struct or remove the `opt.input.is_none()` from the check here. reasoning being the doc string says "/// GGA update period, in seconds. 0 means to never send a GGA" but here it seems like the behavior is if period is 0 and input is empty then don't send anything, if period is 0 and there is an input, then you send cmds as fast as possible because `cmd.after` is set to zero.

Reply:
`--input` disregards most of the other flags, I added some clap stuff to throw and error for mixing them but missed this one. I tweaked the logic to make it more clear, I think. The timing of the messages when you use `--input` is based solely on the file contents, if you don't set `after: x` it will send immediately after the previous one. Maybe it should default to 10 seconds though

Reply:
I see, i don't think it is a good idea to send as many requests as possible, even thought they are serial, so it's not like it is gonna spawn millions of requests... i think it may make sense in this case to derive a default value on the serde deserialization file, but you could also leave it as an option and handle it later.